<div style="text-align: center;">
  <img src="https://github.com/Hack-io-Data/Imagenes/blob/main/01-LogosHackio/logo_naranja@4x.png?raw=true" alt="esquema" />
</div>

# Laboratorio Agrupamiento de Datos

En este laboratorio usaremos el DataFrame creado en la lección anterior resultado de la unión de los conjuntos de datos compartidos contigo. 

**Instrucciones:**

1. Lee cuidadosamente el enunciado de cada ejercicio.

2. Implementa la solución en la celda de código proporcionada.

3. Devuelve los resultados siempre en formato de DataFrame

In [1]:
import pandas as pd 

In [2]:
df_set = pd.read_csv("merged_netflix_data.csv")

In [3]:
df_set

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA        NaN   
3     September 24, 2021          2021  TV-MA        NaN   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R        NaN   
8805    January 11, 2020          2006     PG        NaN   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description                 Title  \
0     As her father nears the end of his life, filmm...  Dick Johnson Is Dead   
1     After crossing paths at a party, a Cape Town t...                   NaN   
2     To protect his family from a powerful drug lor...                   NaN   
3     Feuds, flirtations and toilet talk go down amo...                   NaN   
4     In a city of coaching centers known to train I...                   NaN   
...                                                 ...                   ...   
8802  A political cartoonist, a crime reporter and a...                   NaN   
8803  While living alone in a spooky town, a young g...                   NaN   
8804  Looking to survive in a world taken over by zo...                   NaN   
8805  Dragged from civilian life, a former superhero...                   NaN   
8806  A scrapp

In [4]:
df_set.rename(columns= {"IMDB Score" :"IMDB_Score"} , inplace= True)

#### Ejercicio 1: ¿Cuántos títulos hay por cada tipo de contenido ('Movie' o 'TV Show')?

In [5]:
titulos_agrupados = df_set.groupby("type").size()
print(f"Estos son los titulos que hay por cada tipo de contenido : {titulos_agrupados}")

Estos son los titulos que hay por cada tipo de contenido : type
Movie      6131
TV Show    2676
dtype: int64


#### Ejercicio 2: ¿Cuál es el promedio de puntuación de las películas agrupadas por clasificación de audiencia ('rating')? Interpreta los resultados, ¿hay algo que te llame la atención? Explica por qué. 

In [ ]:
grupo_rating = df_set.groupby("rating")

media_puntuacion = grupo_rating["IMDB_Score"].mean()

print(f"Este es el promedio de las peliculas agrupadas : {media_puntuacion}")




Este es el promedio de las peliculas agrupadas : rating
66 min           NaN
74 min           NaN
84 min           NaN
G           6.500000
NC-17            NaN
NR               NaN
PG          6.800000
PG-13       6.386957
R           6.495745
TV-14       6.177660
TV-G        6.156250
TV-MA       6.097638
TV-PG       6.308772
TV-Y        6.780000
TV-Y7       6.240000
TV-Y7-FV         NaN
UR               NaN
Name: IMDB_Score, dtype: float64


## VALORACION ##

En cuanto a la lectura de los resultados tenemos:
- Errores en los datos: por ejemplo las etiquetas como 66 min, 74 min o 84 min, no parecen clasificacion de audiencia, sino duracion. 
- Falta de datos : existen nulos
- Promedios mas altos : PG (6.8) y TV-Y (6.78) . Tienen mas exito las series para niños de todas las edades (TV-Y) y las que recomiendan supervision de los padres (PG)
- Promedios mas bajos : TV-MA (6.09) y TV-G (6.15). Sin embargo tienen peor calificacion los programas diseñados para mayores de 17 (TV-MA) y las que estan disponibles para todos los publicos (TV-G)

Por un lado tenemos que verificar o ajustar los datos incorrectos (NaN), y la revision de los titulos que se clasifican, por si estuvieramos cometiendo errores de clasificacion por edad o simplemente no estuvieran gustando y tuvieramos que revisar el contenido que estamos ofreciendo a nuestro publico.

#### Ejercicio 3: ¿Cuántos títulos hay por cada país? Ordena los resultados de mayor a menor

In [11]:
titulos_por_pais = df_set.groupby("country").size()

titulos_por_pais_ordenados = titulos_por_pais.sort_values(ascending = False)

titulos_por_pais_ordenados

country
United States          2818
India                   972
United Kingdom          419
Japan                   245
South Korea             199
                       ... 
Venezuela                 1
Venezuela, Colombia       1
, France, Algeria         1
West Germany              1
Zimbabwe                  1
Length: 748, dtype: int64

#### Ejercicio 4: ¿Cuál es la calificación promedio en IMDb de los títulos agrupados por tipo ('Movie' o 'TV Show')?

In [12]:


puntuacion_prom =  df_set.groupby("type")["IMDB_Score"].mean()
print(puntuacion_prom)

type
Movie      6.210916
TV Show         NaN
Name: IMDB_Score, dtype: float64


#### Ejercicio 5: ¿Cuántos títulos se añadieron a Netflix cada año? Para resolver este ejercicio tendrás uqe convertir la columna `date_added` a formato fecha usando el método `pd.to_datetime`. [Aquí](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html) tienes la documentación, leeala con detenimiento antes de usar el método. 

In [13]:
df_set.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description',
       'Title', 'Genre', 'Premiere', 'Runtime', 'IMDB_Score', 'Language',
       'Originales'],
      dtype='object')

In [16]:
print(df_set["date_added"])

0       September 25, 2021
1       September 24, 2021
2       September 24, 2021
3       September 24, 2021
4       September 24, 2021
               ...        
8802     November 20, 2019
8803          July 1, 2019
8804      November 1, 2019
8805      January 11, 2020
8806         March 2, 2019
Name: date_added, Length: 8807, dtype: object


In [17]:
df_set["date_added"] = pd.to_datetime(df_set["date_added"], errors = 'coerce')

In [18]:
df_set = df_set.dropna(subset=["date_added"])

In [20]:
datos_no_validos = df_set[df_set["date_added"].isna()]
print(f"los datos invalidos son : {datos_no_validos}")

los datos invalidos son : Empty DataFrame
Columns: [show_id, type, title, director, cast, country, date_added, release_year, rating, duration, listed_in, description, Title, Genre, Premiere, Runtime, IMDB_Score, Language, Originales]
Index: []


In [21]:
df_set["year_added"] = df_set["date_added"].dt.year

/var/folders/zl/bzzh9yf51f91fkywwtt0dzsc0000gn/T/ipykernel_30961/159905131.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_set["year_added"] = df_set["date_added"].dt.year


In [25]:
titulos_por_anio = df_set.groupby("year_added").size()

In [23]:
titulos_por_anio

year_added
2008       2
2009       2
2010       1
2011      13
2012       3
2013      10
2014      23
2015      73
2016     418
2017    1164
2018    1625
2019    1999
2020    1878
2021    1498
dtype: int64

#### Ejercicio 6: ¿Cuál es el país con más títulos lanzados en el año 2020? Para resolver este ejercicio después del gropuby os puede resultar interesante el método [`idxmax`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html)

In [27]:
df_2020 = df_set[df_set["year_added"] == 2020]
titulos_por_pais = df_2020.groupby("country")["Title"].count()


pais_con_mas_titulos = titulos_por_pais.idxmax()
titulos_totales = titulos_por_pais.max()

print(pais_con_mas_titulos)
print(titulos_totales)

United States
73


#### Ejercicio 7:¿Cuál es el promedio de puntuación en IMDb de los títulos agrupados por país?

In [29]:
df_promedio_IMDB_paises = df_2020.groupby("country")["IMDB_Score"].mean().reset_index()
df_promedio_IMDB_paises

country  IMDB_Score
0                                Argentina        6.42
1                         Argentina, Spain         NaN
2                                Australia         NaN
3    Australia, New Zealand, United States         NaN
4                Australia, United Kingdom         NaN
..                                     ...         ...
244                                Uruguay         NaN
245                     Uruguay, Argentina         NaN
246     Uruguay, Argentina, Germany, Spain         NaN
247                                Vietnam         NaN
248                               Zimbabwe         NaN

[249 rows x 2 columns]

#### Ejercicio 8: ¿Cuántos títulos fueron lanzados en cada año agrupados por tipo ('Movie' o 'TV Show')?

In [30]:
df_titulos_ano_type = df_set.groupby(['year_added', 'type'])['Title'].count().sort_index().reset_index()
df_titulos_ano_type

year_added     type  Title
0         2008    Movie      0
1         2008  TV Show      0
2         2009    Movie      0
3         2010    Movie      0
4         2011    Movie      0
5         2012    Movie      0
6         2013    Movie      0
7         2013  TV Show      0
8         2014    Movie      0
9         2014  TV Show      0
10        2015    Movie     10
11        2015  TV Show      0
12        2016    Movie     28
13        2016  TV Show      0
14        2017    Movie     57
15        2017  TV Show      0
16        2018    Movie     83
17        2018  TV Show      0
18        2019    Movie    107
19        2019  TV Show      0
20        2020    Movie    164
21        2020  TV Show      0
22        2021    Movie     64
23        2021  TV Show      0

#### Ejercicio 9: ¿Cuál es el país con la mayor cantidad de títulos de clasificación 'TV-MA'?

In [31]:
titulos_TV_MA = df_set.groupby("country")["rating"].count().sort_values(ascending=False)
titulos_TV_MA

mas_pais_TV_MA = titulos_TV_MA.idxmax()
total_titulos_TV_MA = titulos_TV_MA.max()
print(mas_pais_TV_MA)
print(total_titulos_TV_MA)

United States
2778


#### Ejercicio 10. ¿Cuál es el promedio de duración de las películas agrupadas por país?

In [32]:
df_promedio_duracion_paises = df_set.groupby("country")["Runtime"].mean().sort_values(ascending=False).reset_index()
df_promedio_duracion_paises

country  Runtime
0          Norway, Iceland, United States    144.0
1      United Kingdom, Hungary, Australia    140.0
2                          Spain, Germany    139.0
3                 Cambodia, United States    136.0
4    United States, United Kingdom, India    130.0
..                                    ...      ...
740                             Venezuela      NaN
741                   Venezuela, Colombia      NaN
742                               Vietnam      NaN
743                          West Germany      NaN
744                              Zimbabwe      NaN

[745 rows x 2 columns]

#### Ejercicio 11: ¿Cuál es el número total de títulos por país y clasificación de audiencia ('rating')?

In [33]:
df_paises_titulos_rating = df_set.groupby(['country', 'rating'])['Title'].count().sort_values(ascending=False).reset_index()
df_paises_titulos_rating

country rating  Title
0                                United States  TV-MA    121
1                                United States  TV-14     47
2                                United States  TV-PG     35
3                                United States      R     34
4                                        India  TV-MA     19
...                                        ...    ...    ...
1160             United States, United Kingdom   TV-Y      0
1161  United States, United Kingdom, Australia  PG-13      0
1162  United States, United Kingdom, Australia  TV-Y7      0
1163     United States, United Kingdom, Canada     PG      0
1164                Argentina, Uruguay, Serbia  TV-14      0

[1165 rows x 3 columns]

#### Ejercicio 12: ¿Cuántos títulos fueron lanzados antes del año 2000 agrupados por país?

In [36]:
df_titulos_previo_2020 = df_set[df_set["year_added"] < 2000]
titulos_previo_2000 = df_titulos_previo_2020.groupby("country")["Title"].count()
titulos_previo_2000

Series([], Name: Title, dtype: int64)

#### Ejercicio 13: ¿Cuál es la puntuación promedio en IMDb de los títulos lanzados en 2021 agrupados por país?

In [37]:
df_titulos_previo_2021 = df_set[df_set["year_added"] < 2021]
df_titulos_previo_2021_A = df_titulos_previo_2021.groupby('country')['IMDB_Score'].mean().sort_values(ascending=False).reset_index()

df_titulos_previo_2021_A

country  IMDB_Score
0    United Kingdom, Ukraine, United States         8.4
1                     Spain, United Kingdom         8.2
2             United States, United Kingdom         8.2
3      United States, United Kingdom, India         7.8
4                  Australia, United States         7.6
..                                      ...         ...
654                               Venezuela         NaN
655                     Venezuela, Colombia         NaN
656                                 Vietnam         NaN
657                            West Germany         NaN
658                                Zimbabwe         NaN

[659 rows x 2 columns]

#### Ejercicio 14: ¿Cuántos títulos fueron dirigidos por cada director? Ordena los resultados de mayor a menor

In [38]:
df_titulos_por_director = df_set.groupby("director")["Title"].count().sort_values(ascending=False).reset_index()
df_titulos_por_director

director  Title
0                        McG      3
1              Blair Simmons      2
2     Bonni Cohen, Jon Shenk      2
3                Brian Oakes      2
4               Zak Hilditch      2
...                      ...    ...
4519             Çagan Irmak      0
4520        Ísold Uggadóttir      0
4521     Óskar Thór Axelsson      0
4522        Ömer Faruk Sorak      0
4523            Aaron Lieber      0

[4524 rows x 2 columns]

#### Ejercicio 15: ¿Cuál es el lenguaje más frecuente en los títulos lanzados en los últimos 5 años agrupado por país?

In [39]:
ultimos_5_anios = df_set[df_set["year_added"] >= 2019]
top_idiomas = ultimos_5_anios.groupby(["country", "Language"])["Title"].count()
top_idiomas

country                                 Language  
Argentina                               Spanish       5
Austria                                 German        1
Belgium, Netherlands                    Dutch         1
Brazil                                  Portuguese    6
Canada                                  English       3
                                                     ..
United States, Mexico                   Spanish       1
United States, United Kingdom           English       1
United States, United Kingdom, Canada   English       1
United States, United Kingdom, Germany  English       1
United States, United Kingdom, India    English       1
Name: Title, Length: 76, dtype: int64